File: backend/application/use_cases/eliminar_tarea.py
# ═══════════════════════════════════════════════════════════════════════════════
# CAPA: APPLICATION — Use Case: EliminarTarea
# ═══════════════════════════════════════════════════════════════════════════════

from domain.entities.tarea import Tarea
from application.ports.itarea_repository import ITareaRepository


class EliminarTarea:
    """
    Use Case: Eliminar una Tarea por su ID.
    """

    def __init__(self, repo: ITareaRepository) -> None:
        self.repo = repo

    def execute(self, id: str) -> None:
        """
        Elimina la Tarea con el ID dado.

        -> None significa que no devuelve ningún valor.
        Si la Tarea no existe, lanza ValueError (el router lo convierte en HTTP 404).
        """

        # ── Delegar la eliminación al repositorio ─────────────────────────────
        fue_eliminada: bool = self.repo.eliminar(id)
        # bool = True si se eliminó, False si el ID no existía.

        if not fue_eliminada:
            # "not fue_eliminada" = "if fue_eliminada is False"
            raise ValueError(f"No se encontró una Tarea con id '{id}'.")
            # El Adapter (router) captura esta excepción y devuelve HTTP 404.

        # Si llegamos acá sin excepción, la eliminación fue exitosa.
        # El router devolverá HTTP 204 (No Content) — éxito sin cuerpo de respuesta.

File: frontend/src/application/useCases/Tareas/EliminarTarea.ts
// ═══════════════════════════════════════════════════════════════════════════════
// CAPA: APPLICATION — Use Case: EliminarTarea
//
// Posición en la cadena de dependencias:
//   Adapter (mutationFn) → EliminarTarea.execute(id) → repo.eliminar(id) → HTTP DELETE
//
// Orquesta la eliminación: delega al repo. No hay validación de Domain en este caso
// porque el Domain no tiene reglas de negocio para QUIÉN puede eliminar (en este CRUD).
// En un sistema real, aquí verificaría permisos: ¿puede este rol eliminar?
//
// Regla de dependencias (Clean Architecture — Ley de Dependencia):
//   ✅ Puede importar: application/ports
//   ❌ NO puede importar: fetch, React, TareaRepositoryImpl, nada de Infrastructure
//
// 🔍 DevTools — cómo observar este archivo en acción:
//   Network > Fetch/XHR: DELETE /api/v1/tareas/{id} → status 204 No Content
//   La columna 'Size' en Network tab = 0 B (sin body — HTTP 204 no tiene cuerpo).
//   React Query DevTools: después del DELETE exitoso → query ['tareas', *] pasa a 'stale'.
// ═══════════════════════════════════════════════════════════════════════════════

import type { ITareaRepository } from '@/application/ports/ITareaRepository'


export class EliminarTarea {
  private repo: ITareaRepository

  constructor(repo: ITareaRepository) {
    this.repo = repo
  }

  async execute(id: string): Promise<void> {
    // "Promise<void>" = operación async que no devuelve ningún valor al terminar.
    // "void" en TypeScript = ausencia de valor. No es null, no es undefined — es nada.
    //
    // Iron Law 2: TODAS las escrituras pasan por UseCase.
    // El componente llama al hook → hook llama a execute() → execute() llama al repo.
    // NUNCA: componente → repo directamente.

    // 🔍 Network: DELETE /api/v1/tareas/{id} → status 204 sin body.
    // Si ves status 404 → la tarea no existe → el repo lanzará un Error.
    await this.repo.eliminar(id)
    // Si la tarea no existe, el repo lanza un Error.
    // El hook recibirá ese Error en "onError" y podrá mostrarlo en la UI.
  }
}
